src/js/entities-service/entities/clinicians.js

```diff
@@ -0,0 +1,22 @@
+import BaseEntity from 'js/base/entity-service';
+import { _Model, Model, Collection } from './entities/clinicians';
+
+const Entity = BaseEntity.extend({
+  Entity: { _Model, Model, Collection },
+  radioRequests: {
+    'clinicians:model': 'getModel',
+    'clinicians:collection': 'getCollection',
+    'fetch:clinician:model': 'fetchClinician',
+  },
+  fetchClinician(id) {
+    const include = [
+      'role',
```

Comment:
@luketlancaster @paulfalgout the care ops backend does not support includes, and I don't really want to add support for them because it makes for some really questionable performance.

From my perspective, the JS models should use the `links` attached to responses to generate new promises to fetch related data on demand. For instance (syntax might be different):

```js
let roles = await model.clinicians.get(clinicianId).roles();
```

Reply:
just to be clear here.. you're suggesting when we have a list of things that have relationships.. like the action list that relates to the patient and clinician that we should make (number of results * number of relationships + 1) requests to show the list?

Reply:
Multiple requests in parallel, yes. Something like https://www.npmjs.com/package/swagger-fluent would probably help.

Reply:
well we could setup a node server to do the packaging.  that `swagger-fluent` lib is for node.  _technically_ it can be run in the browser but it's over 3M, and the author of `request`, it's dependency suggests, it shouldn't be used in the browser.  To my knowledge most frontend solutions have been working to reduce roundtrips because in addition to response time there's latency and concurrency limitations with browsers.  Perhaps we could find some other resources that suggest this is a good path or a library with more than one contributor and a user-base that does it this way?

Reply:
Also even in your example it wouldn't be 5 async requests.  It'd be 1 request vs 1 request + 4 async requests (assuming only 1 level of relationships)

Reply:
https://hackernoon.com/5-small-steps-to-a-better-web-api-6fa9698178a6
https://softwareengineering.stackexchange.com/a/305875

I assume that the top level item will probably already have a local reference. For instance:

- when the clinician first accesses the page:
  - parallel (clinician, organization)
  - parallel (groups, roles)
- when the clinician views a group:
  - await patients
- when the clinician views a patient:
  - (patient already loaded)
  - parallel (groups, fields, actions, ...)

Reply:
Also, HTTP/2 support when available makes parallel async *much* faster than a single large request. It also utilizes lambda better, as lambda will automatically scale the server up and down to deal with higher traffic. Lambda will (probably) not scale up as quickly for a small number of requests that take a long time to complete.

Reply:
I think it'd be more like:
- when the clinician first accesses the page:
  - parallel (clinician, organizations)
  - parallel (clinician.groups, clinician.roles)
  - parallel (each group's group.clinicians needed across the app for various functionality) - we'll be loading any clinician in multiple groups multiple times.
- when the clinician views a group:
  - (group already loaded via clinician.groups)
  - await group.patients
- when the clinician views a patient:
  - ~~(patient already loaded via group.patients)~~ since we don't know for sure if the patient data is still valid, load it again
  - parallel (patient.groups, patient.fields, patient.actions, ...)

Reply:
If loading all of the organization groups, do:

- ~~parallel (clinician.groups, clinician.roles)~~ parallel (organization.groups, organization.roles)

Reply:
Oh and this article (which I was hunting for and finally found):

https://apisyouwonthate.com/blog/lets-stop-building-apis-around-a-network-hack